mods/flowers/init.lua

```diff
@@ -139,32 +139,36 @@ function flowers.flower_spread(pos, node)
 
 	local pos0 = vector.subtract(pos, 4)
 	local pos1 = vector.add(pos, 4)
-	if #minetest.find_nodes_in_area(pos0, pos1, "group:flora") > 3 then
+	-- A limit of 11 in a 9*9 area, and 3 added at once, results in an average
+	-- maximum density of 13, this matches maximum mapgen flower density.
+	if #minetest.find_nodes_in_area(pos0, pos1, "group:flora") > 11 then
 		return
 	end
 
 	local soils = minetest.find_nodes_in_area_under_air(
 		pos0, pos1, "group:soil")
-	if #soils > 0 then
-		local seedling = soils[math.random(#soils)]
-		local seedling_above =
-			{x = seedling.x, y = seedling.y + 1, z = seedling.z}
-		light = minetest.get_node_light(seedling_above)
-		if not light or light < 13 or
-				-- Desert sand is in the soil group
-				minetest.get_node(seedling).name == "default:desert_sand" then
-			return
-		end
+	local num_soils = #soils
+	if num_soils >= 1 then
+		for si = 1, math.min(3, num_soils) do
+			local soil = soils[math.random(num_soils)]
+			local soil_above = {x = soil.x, y = soil.y + 1, z = soil.z}
+			light = minetest.get_node_light(soil_above)
+			if not light or light < 13 or
+					-- Desert sand is in the soil group
+					minetest.get_node(soil).name == "default:desert_sand" then
+				return
```

Comment:
Don't return here. Leave this case empty and set the node in `else`

Reply:
Ok.

Reply:
Ugh a return in a loop, sorry, big mistake.

Reply:
Fixed, testing.